crates/astria-proto/src/native/sequencer.rs

```diff
@@ -0,0 +1,326 @@
+    ///     Address,
+    ///     ADDRESS_LEN,
+    /// };
+    /// let address = Address::new([42; ADDRESS_LEN]);
+    /// let buffer = address.to_vec();
+    /// assert_eq!(address.as_ref(), &buffer);
+    /// ```
+    #[must_use]
+    pub fn to_vec(self) -> Vec<u8> {
+        self.0.to_vec()
+    }
+}
+
+impl AsRef<[u8]> for Address {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl From<[u8; ADDRESS_LEN]> for Address {
+    fn from(inner: [u8; ADDRESS_LEN]) -> Self {
+        Self(inner)
+    }
+}
+
+impl Display for Address {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use hex::ToHex as _;
+        f.write_str(&self.encode_hex::<String>())
+    }
+}
+
+impl v1alpha1::BalanceResponse {
+    /// Converts an astria native [`BalanceResponse`] to a
+    /// protobuf [`v1alpha1::BalanceResponse`].
+    #[must_use]
+    pub fn from_native(native: BalanceResponse) -> Self {
+        let BalanceResponse {
+            account,
+            height,
+            balance,
+        } = native;
+        Self {
+            account: account.0.to_vec(),
+            height,
+            balance: Some(balance.into()),
+        }
+    }
+
+    /// Converts a protobuf [`v1alpha1::BalanceResponse`] to an astria
+    /// native [`BalanceResponse`].
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the account buffer could not be converted to an [`Address`] because
+    /// it was not 20 bytes long.
+    pub fn into_native(self) -> Result<BalanceResponse, IncorrectAddressLength> {
+        BalanceResponse::from_proto(self)
+    }
+
+    /// Converts a protobuf [`v1alpha1::BalanceResponse`] to an astria
+    /// native [`BalanceResponse`] by allocating a new [`v1alpha::BalanceResponse`].
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the account buffer could not be converted to an [`Address`] because
+    /// it was not 20 bytes long.
+    pub fn to_native(&self) -> Result<BalanceResponse, IncorrectAddressLength> {
+        self.clone().into_native()
+    }
+}
+
+/// The sequencer response to a balance request for a given account at a given height.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct BalanceResponse {
+    pub account: Address,
+    pub height: u64,
+    pub balance: u128,
+}
+
+impl BalanceResponse {
+    /// Converts a protobuf [`v1alpha1::BalanceResponse`] to an astria
+    /// native [`BalanceResponse`].
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the account buffer could not be converted to an [`Address`] because
+    /// it was not 20 bytes long.
+    pub fn from_proto(proto: v1alpha1::BalanceResponse) -> Result<Self, IncorrectAddressLength> {
+        let v1alpha1::BalanceResponse {
+            account,
+            height,
+            balance,
+        } = proto;
+        Ok(Self {
+            account: Address::try_from_slice(&account)?,
+            height,
+            balance: balance.map_or(0, Into::into),
```

Comment:
should this error if balance is `None`?

Reply:
I think this is also kinda correct here? A balance of none doesn't really make sense.